### ISSUE TITLE
data race flake in TestListener and TestControlListener

### ISSUE THREAD

**Issue by mvdan** [TykTechnologies/tyk#1305]
For `TestListener`:

```
==================
WARNING: DATA RACE
Write at 0x00c42bbb9590 by goroutine 124:
  runtime.mapassign_faststr()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/runtime/hashmap_fast.go:598 +0x0
  github.com/TykTechnologies/tyk.TestMain.func2()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:179 +0x1d7

Previous read at 0x00c42bbb9590 by goroutine 121:
  [failed to restore the stack]

Goroutine 124 (running) created at:
  github.com/TykTechnologies/tyk.TestMain()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:169 +0x4b8
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd

Goroutine 121 (running) created at:
  github.com/TykTechnologies/tyk.TestMain()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:166 +0x488
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd
==================
time="Nov 21 05:12:01" level=info msg="Preparing new router" 
time="Nov 21 05:12:01" level=info msg="Initialising Tyk REST API Endpoints" 
time="Nov 21 05:12:01" level=info msg="Loading API configurations." 
time="Nov 21 05:12:01" level=info msg="Loading uptime tests..." 
time="Nov 21 05:12:01" level=info msg="Initialised API Definitions" 
time="Nov 21 05:12:01" level=info msg="API reload complete" 
time="Nov 21 05:12:01" level=info msg="Initiating coprocess reload" 
==================
WARNING: DATA RACE
Write at 0x000001932430 by goroutine 420:
  github.com/TykTechnologies/tyk.(*JSVM).Init()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/mw_js_plugin.go:322 +0x365
  github.com/TykTechnologies/tyk.setupGlobals()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/main.go:139 +0x1081
  github.com/TykTechnologies/tyk.testHttp()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:828 +0xca0
  github.com/TykTechnologies/tyk.TestListener()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:933 +0xf4
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:746 +0x16c

Previous write at 0x000001932430 by goroutine 121:
  [failed to restore the stack]

Goroutine 420 (running) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:789 +0x568
  testing.runTests.func1()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:1004 +0xa7
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:746 +0x16c
  testing.runTests()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:1002 +0x521
  testing.(*M).Run()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:921 +0x206
  github.com/TykTechnologies/tyk.TestMain()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:186 +0x4c9
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd

Goroutine 121 (running) created at:
  github.com/TykTechnologies/tyk.TestMain()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:166 +0x488
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd
==================
==================
WARNING: DATA RACE
Write at 0x000001932438 by goroutine 420:
  github.com/TykTechnologies/tyk.(*JSVM).Init()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/mw_js_plugin.go:323 +0x393
  github.com/TykTechnologies/tyk.setupGlobals()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/main.go:139 +0x1081
  github.com/TykTechnologies/tyk.testHttp()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:828 +0xca0
  github.com/TykTechnologies/tyk.TestListener()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:933 +0xf4
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:746 +0x16c

Previous write at 0x000001932438 by goroutine 121:
  [failed to restore the stack]

Goroutine 420 (running) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:789 +0x568
  testing.runTests.func1()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:1004 +0xa7
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:746 +0x16c
  testing.runTests()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:1002 +0x521
  testing.(*M).Run()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:921 +0x206
  github.com/TykTechnologies/tyk.TestMain()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:186 +0x4c9
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd

Goroutine 121 (running) created at:
  github.com/TykTechnologies/tyk.TestMain()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:166 +0x488
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd
==================
==================
WARNING: DATA RACE
Write at 0x000001932440 by goroutine 420:
  github.com/TykTechnologies/tyk.(*JSVM).Init()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/mw_js_plugin.go:324 +0x3d5
  github.com/TykTechnologies/tyk.setupGlobals()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/main.go:139 +0x1081
  github.com/TykTechnologies/tyk.testHttp()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:828 +0xca0
  github.com/TykTechnologies/tyk.TestListener()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:933 +0xf4
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:746 +0x16c

Previous write at 0x000001932440 by goroutine 121:
  [failed to restore the stack]

Goroutine 420 (running) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:789 +0x568
  testing.runTests.func1()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:1004 +0xa7
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:746 +0x16c
  testing.runTests()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:1002 +0x521
  testing.(*M).Run()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:921 +0x206
  github.com/TykTechnologies/tyk.TestMain()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:186 +0x4c9
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd

Goroutine 121 (running) created at:
  github.com/TykTechnologies/tyk.TestMain()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:166 +0x488
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd
==================
```

For `TestControlListener`:

```
==================
WARNING: DATA RACE
Write at 0x00c423876780 by goroutine 124:
  runtime.mapassign_faststr()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/runtime/hashmap_fast.go:598 +0x0
  github.com/TykTechnologies/tyk.TestMain.func2()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:179 +0x1d7

Previous read at 0x00c423876780 by goroutine 135:
  github.com/TykTechnologies/tyk.doReload()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/main.go:602 +0x64a
  github.com/TykTechnologies/tyk.TestControlListener()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:951 +0xb2
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:746 +0x16c

Goroutine 124 (running) created at:
  github.com/TykTechnologies/tyk.TestMain()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:169 +0x4b8
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd

Goroutine 135 (running) created at:
  testing.(*T).Run()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:789 +0x568
  testing.runTests.func1()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:1004 +0xa7
  testing.tRunner()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:746 +0x16c
  testing.runTests()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:1002 +0x521
  testing.(*M).Run()
      /home/travis/.gimme/versions/go1.9.2.linux.amd64/src/testing/testing.go:921 +0x206
  github.com/TykTechnologies/tyk.TestMain()
      /home/travis/gopath/src/github.com/TykTechnologies/tyk/gateway_test.go:186 +0x4c9
  main.main()
      github.com/TykTechnologies/tyk/_test/_testmain.go:368 +0x1cd
==================
```

**Comment by mvdan** [comment]
Perhaps these are all simply because of the PR above.

**Comment by mvdan** [comment]
Not because of the PR - being fixed in #1313.
